### ISSUE TITLE
Detekt's 2.0 architecture

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#2680]
There are PRs and issues talking about how to arrange the code in the project for the 2.0. Some PRs moving classes and creating aliases to don't break the compatibility for "1.0" but preparing the 2.0 version.

I'm not saying that we are near to create the 2.0 version. I think that there is not a big reason right now to release a major release with breaking changes. But, anyway, we can start talking about how should be the architecture of detekt. Right now I'm mainly thinking about the gradle modules, but we can talk about other topics too.

Ideas:
- Which modules we do need? Which is the responsability for each of them?
- How do we make easy to understand what's our public API and what's our private api?

For example. Right now `:detekt-gradle-plugin` depends on `:detekt-cli`. I think that we should break that dependency. Both should be *clients* of/depend on the same module that exposes all the functionality.

Should we expose all our public api in a single module or split it? Rules/post processors/metrics?/something else?

Could we hide the implementation details from that modules? I mean, could we make does modules without depend on our "core"? This way we reduce the accidental use of our private API.

Other option to avoid this problem is to create a custom rule so the third parties can use that rule and warn them if they are using our private API.

**Comment by arturbosch** [comment]
Nice idea to start collecting/discussing further architecture changes.

A `:detekt-tooling-api` could introduce interfaces/service loaders for embedding detekt.
Some ideas:
```kotlin
interface DetektCliWrapper {
    fun execute(args: Array<String>): ExitCode
}
```

```kotlin
interface RunnerProvider {
    fun load(args: Array<String>, out: Appendable, error: Appendable): Runner
}
```

```kotlin
interface DetektFacadeProvider {
    fun load(settings: ProcessingSettings, classloader: ClassLoader?): DetektFacade
}
```

```kotlin
interface DetektFacade {
    fun run(files: Collection<KtFile>, context: BindingContext): Result
}
```

**Comment by schalkms** [comment]
I also feel that this issue is a good starting point for collecting ideas. Thanks for that! @BraisGabin 

> For example. Right now :detekt-gradle-plugin depends on :detekt-cli. I think that we should break that dependency. Both should be clients of/depend on the same module that exposes all the functionality.

That's the biggest pain point. For me this has the highest priority.

The rest is IMHO nice to have. As a user, I'd like to have a **long term stable** detekt API. Users don't want to change their plugins and 3rd party rules every few months.

Regarding the visibility of the API functions, I think Artur has done a good job here. Since detekt is extensible in many ways, many interfaces are exposed.

**Comment by BraisGabin** [comment]
> That's the biggest pain point. For me this has the highest priority.

Probably we can address this point without any breaking change. But I don't know if it's worth it to make a refactor like that forcing us to compatibility

> The rest is IMHO nice to have. As a user, I'd like to have a long term stable detekt API. Users don't want to change their plugins and 3rd party rules every few months.

I agree. We should wait to have a really good reason. Or multiple medium reasons.

> Regarding the visibility of the API functions, I think Artur has done a good job here. Since detekt is extensible in many ways, many interfaces are exposed.

I agree. Detekt is really extensible. My point is that we should create clear separations between public api and private api so we can refactor all we want easier.

**Comment by arturbosch** [comment]
When further investing into metrics, a general `Metric` interface is desirable.
We have classes `Metric` and `ProjectMetric` which has basically has the same properties.
The only difference is the subject the metric belongs, a project or a construct in the code.

```kotlin
interface Metric {
    val type: String,
    val value: Int,
    val component: Component,
    val threshold: Int? = null,
    val isDouble: Boolean = false,
    val conversionFactor: Int = DEFAULT_FLOAT_CONVERSION_FACTOR
}
```

```kotlin
enum class Component {
    FUNCTION,
    METHOD,
    CLASS,
    FILE,
    DIRECTORY,
    PROJECT
}
```

**Comment by arturbosch** [comment]
We have two concepts to suppress issues: `Baseline` and `@Suppress` annotations.
These are cool concepts which allow to introduce detekt and it's rules gradually.
However these concepts are implemented as "special cases". The information if an issue was suppressed gets lost, the user just sees the non suppressed findings.
Tools may want to import an issue in a kind of database for evolutionary analysis.

We could mark an issue as `isSuppressed` and give a `suppressionKind`:
```kotlin
enum class SuppressionKind {
    ANNOTATION,
    COMMENT,
    BASELINE
}
```

Filtering issues could be generalized as:

```kotlin
interface ResultMapping {
    val priority: Int = Int.MAX
    fun map(result: Result): Result
}
```

Some kind of `fun onStart(startingResult: Result)` could be used if the unfiltered result collection is needed.

**Comment by arturbosch** [comment]
To minimize changes across different modules when upgrading the kotlin compiler dependency, I propose to introduce a new module `detekt-kotlinc-utils` / `detekt-psi-utils` / `detekt-compiler` / `detekt-parser` or something.

This module may include `KotlinEnvironmentUtils`, `KtCompiler`, `KtTreeCompiler`, `DetektPomModel` etc.

**Comment by arturbosch** [comment]
Many classes/functions inside `detekt-cli` can be seen as core features/tooling interfaces of detekt because they are used from the Gradle/Sonar/IntelliJ plugin to bootstrap detekt e.g. `BaselineFacade`, `CliArgs`, `Runner`, `fun CliArgs.loadConfiguration(): Config`.

As @BraisGabin and @schalkms have pointed out a core module which can be consumed by the cli and gradle plugin is desirable.
Supporting additional plugins/tools through a `tooling-api` / interface (https://github.com/detekt/detekt/issues/2680#issuecomment-628875163) would allow to minimize the dependencies on above core classes.

**Comment by BraisGabin** [comment]
> When further investing into metrics, a general `Metric` interface is desirable.
> We have classes `Metric` and `ProjectMetric` which has the same properties.
> The only difference is the subject the metric belongs, a project or a construct in the code.
> 
> ```kotlin
> interface Metric {
>     val type: String,
>     val value: Int,
>     val component: Component,
>     val threshold: Int? = null,
>     val isDouble: Boolean = false,
>     val conversionFactor: Int = DEFAULT_FLOAT_CONVERSION_FACTOR
> }
> ```
> 
> ```kotlin
> enum class Component {
>     FUNCTION,
>     METHOD,
>     CLASS,
>     FILE,
>     DIRECTORY,
>     PROJECT
> }
> ```

Open questions (I don't know the answers):

- Do we need type and Component?
- Does `Component` allow any case scenario for 3rd parties? Should it?

**Comment by arturbosch** [comment]
> > When further investing into metrics, a general `Metric` interface is desirable.
> > We have classes `Metric` and `ProjectMetric` which has the same properties.
> > The only difference is the subject the metric belongs, a project or a construct in the code.
> > ```kotlin
> > interface Metric {
> >     val type: String,
> >     val value: Int,
> >     val component: Component,
> >     val threshold: Int? = null,
> >     val isDouble: Boolean = false,
> >     val conversionFactor: Int = DEFAULT_FLOAT_CONVERSION_FACTOR
> > }
> > ```
> > 
> > 
> > ```kotlin
> > enum class Component {
> >     FUNCTION,
> >     METHOD,
> >     CLASS,
> >     FILE,
> >     DIRECTORY,
> >     PROJECT
> > }
> > ```
> 
> Open questions (I don't know the answers):
> 
> * Do we need type and Component?

It's basically the name of the metric.
Maybe we should not name it type but `id` and add `displayName`.

> * Does `Component` allow any case scenario for 3rd parties? Should it?

We could use it to identify `ProjectMetric`s and generate automatically a complexity report for it.
Third party tools can use it to track the evolution of a metric for example.
So maybe we should save an `Entity` for a metric to.

**Comment by BraisGabin** [comment]
We should use more immutable classes. For example `Detektion`. I think that all our public api should use immutable classes.

And then, even the private api should be immutable by default. I don't know all the cases and we can have good reasons to have some mutable ones, but in general I think that immutable is less error prune.

**Comment by arturbosch** [comment]
`RuleSetProvider` should extend `Extension`:

```kotlin
interface RuleSetProvider : Extension {

    /**
     * Every rule set must be pre-configured with an ID to validate if this rule set
     * must be created for current analysis.
     */
    val ruleSetId: String

    /**
     * This function must be implemented to provide custom rule sets.
     * Make sure to pass the configuration to each rule to allow rules
     * to be self configurable.
     */
    fun rules(config: Config): Collection<Rule>
}
```

**Comment by arturbosch** [comment]
`DefaultConfig.RESOURCE_NAME` can be made internal/private after we remove the public reference in cli module.

```kotlin
object DefaultConfig {

    const val RESOURCE_NAME = "default-detekt-config.yml"
```

**Comment by BraisGabin** [comment]
We should unify the package names too.

**Comment by arturbosch** [comment]
We should think about hiding console-reports behind a `verbose` flag and hide all `FileProcessListener`'s. 
They are mainly used by extension authors or the maintainers for implementing metric or rule stuff.
`OutputReport`'s can be configured via cli flags or gradle properties, so we should remove them from the config.

I cross-reference this to #2833

**Comment by BraisGabin** [comment]
What do you think about removing `Config` as a property in the rules? A rule should not know about `Config`. It should receive by constructor the configurations that it needs. For example, instead of:

```kotlin
class ForbiddenImport(config: Config = Config.empty) : Rule(config) {

    private val forbiddenImports = valueOrDefaultCommaSeparated(IMPORTS, emptyList())
        .distinct()
        .map { it.simplePatternToRegex() }

    private val forbiddenPatterns: Regex = Regex(valueOrDefault(FORBIDDEN_PATTERNS, ""))
```

do this:

```kotlin
class ForbiddenImport(
    private val forbiddenImports: List<String>,
    private val forbiddenPatterns: Regex 
) : Rule() {
```

And if we want to handle default values we could do things like this:

```kotlinclass ForbiddenImport(
    forbiddenImports: List<String>?,
    forbiddenPatterns: Regex?
) : Rule() {

    private val forbiddenImports: List<String> = forbiddenImports ?: emptyList()
    private val forbiddenPatterns: Regex? = forbiddenPatterns ?: "".toRegex()
```

But this change is even bigger than this. Other things that we are getting from the configuration:

- `excludes`/`includes`: This should be handled by `:core`, the rule should not know if it should run over a file or not.
- `active`: A disabled rule is the same as no rule. We don't need this parameter in the rule. We can handle this in `:core` or `RuleSet`. If the rule is disabled we don't instanciate it.
- `aliases`: We use the aliases to Suppress issues. This is a tricky one. We support the user to add alias to any rule so this should be handled by `:core` but we should also allow the rule writer to provide it's own aliases so that should be in the rule itself. My idea here is to move `defaultRuleIdAliases` to `BaseRule` so `:core` can access it and take the correct decission.
  - There is a down side with this solution. Right now we have some tests to ensure that `@Suppress` works as intended in some rules. This have sense because it depends on the `KtElement` that we provide in `Finding`. And we have these tests because we had problems there... how are we going to test this things if we move this logic to `:core`?
- `ruleSetId`: Right now this is a nullable property. I think that we should make this a non-null constructor property. When you crate any rule you should provide it.

I think that this will help to implement #2581 and it will remove a lot of complexity from `Rule`.

**Comment by arturbosch** [comment]
> What do you think about removing `Config` as a property in the rules? A rule should not know about `Config`. It should receive by constructor the configurations that it needs. For example, instead of:
> 
> ```kotlin
> class ForbiddenImport(config: Config = Config.empty) : Rule(config) {
> 
>     private val forbiddenImports = valueOrDefaultCommaSeparated(IMPORTS, emptyList())
>         .distinct()
>         .map { it.simplePatternToRegex() }
> 
>     private val forbiddenPatterns: Regex = Regex(valueOrDefault(FORBIDDEN_PATTERNS, ""))
> ```
> 
> do this:
> 
> ```kotlin
> class ForbiddenImport(
>     private val forbiddenImports: List<String>,
>     private val forbiddenPatterns: Regex 
> ) : Rule() {
> ```
> 
> And if we want to handle default values we could do things like this:
> 
> ```
>     forbiddenImports: List<String>?,
>     forbiddenPatterns: Regex?
> ) : Rule() {
> 
>     private val forbiddenImports: List<String> = forbiddenImports ?: emptyList()
>     private val forbiddenPatterns: Regex? = forbiddenPatterns ?: "".toRegex()
> ```
> 
> Or implement a factory.
> 
> But this change is even bigger than this. Other things that we are getting from the configuration:
> 
> * `excludes`/`includes`: This should be handled by `:core`, the rule should not know if it should run over a file or not.
> * `active`: A disabled rule is the same as no rule. We don't need this parameter in the rule. We can handle this in `:core` or `RuleSet`. If the rule is disabled we don't instanciate it.
> * `aliases`: We use the aliases to Suppress issues. This is a tricky one. We support the user to add alias to any rule so this should be handled by `:core` but we should also allow the rule writer to provide it's own aliases so that should be in the rule itself. My idea here is to move `defaultRuleIdAliases` to `BaseRule` so `:core` can access it and take the correct decission.
>   
>   * There is a down side with this solution. Right now we have some tests to ensure that `@Suppress` works as intended in some rules. This have sense because it depends on the `KtElement` that we provide in `Finding`. And we have these tests because we had problems there... how are we going to test this things if we move this logic to `:core`?
> * `ruleSetId`: Right now this is a nullable property. I think that we should make this a non-null constructor property. When you crate any rule you should provide it.
> 
> I think that this will help to implement #2581 and it will remove a lot of complexity from `Rule`.

Not sure yet about removing the whole Config from a Rule because we just have to move it to the `RuleSetProvider` to configure the rule.

I agree with removing the whole 

filter, aliases, active and even suppressing part from the Rule and move it to core.

```kotlin
    open val filters: PathFilters? by lazy(LazyThreadSafetyMode.NONE) {
        createPathFilters()
    }

    override fun visitCondition(root: KtFile): Boolean =
        active &&
            shouldRunOnGivenFile(root) &&
            !root.isSuppressedBy(ruleId, aliases, ruleSetId)

    private fun shouldRunOnGivenFile(root: KtFile) =
        filters?.isIgnored(root.absolutePath())?.not() ?: true
```

We already have a `ruleId` to `ruleSetId` mapping in core and can safely remove `ruleSetId` from `Rule` and do the suppressing in `core`.

**Comment by arturbosch** [comment]
> > > When further investing into metrics, a general `Metric` interface is desirable.
> > > We have classes `Metric` and `ProjectMetric` which has the same properties.
> > > The only difference is the subject the metric belongs, a project or a construct in the code.
> > > ```kotlin
> > > interface Metric {
> > >     val type: String,
> > >     val value: Int,
> > >     val component: Component,
> > >     val threshold: Int? = null,
> > >     val isDouble: Boolean = false,
> > >     val conversionFactor: Int = DEFAULT_FLOAT_CONVERSION_FACTOR
> > > }
> > > ```
> > > 
> > > 
> > > ```kotlin
> > > enum class Component {
> > >     FUNCTION,
> > >     METHOD,
> > >     CLASS,
> > >     FILE,
> > >     DIRECTORY,
> > >     PROJECT
> > > }
> > > ```
> > 
> > 
> > Open questions (I don't know the answers):
> > 
> > * Do we need type and Component?
> 
> It's basically the name of the metric.
> Maybe we should not name it type but `id` and add `displayName`.
> 
> > * Does `Component` allow any case scenario for 3rd parties? Should it?
> 
> We could use it to identify `ProjectMetric`s and generate automatically a complexity report for it.
> Third party tools can use it to track the evolution of a metric for example (visualization of metrics/findings).
> So maybe we should save an `Entity` for a metric to.

Another idea with metrics:

Metrics could be used to define repeatable `MetricViolation` `Rule`s and implement feature request #1224.
The idea of the issue is to have for example `LargeClass` rule enabled for all classes, have one class suppressed and define a new threshold for the suppressed class.

```yaml
metrics:
  ...
  MySuppressedGodClass:
    include: "full/path/to/class.kt"
    metrics:
      - loc > 1000
      - cognitiveComplexity > 50 
```

Another way would be to implement a custom `ReportingExtension` which adds the suppressed Finding again to the result.

**Comment by arturbosch** [comment]
For 2.0 we should probably use yaml for the baseline - #2892.
We already use Yaml for the configuration. The baseline can be easily represented in yaml format.

**Comment by schalkms** [comment]
@arturbosch I have an idea.
I would prefer to have an internal DSL for the configuration over yaml. The DSL should ofc be written in Kotlin. 😉
What do you think?

**Comment by BraisGabin** [comment]
> @arturbosch I have an idea.
> I would prefer to have an internal DSL for the configuration over yaml. The DSL should ofc be written in Kotlin. 😉
> What do you think?

Type checking and autocomplete from the ide... That sounds great to reduce errors and increase the discovery! I really like this idea.

**Comment by arturbosch** [comment]
> @arturbosch I have an idea.
> I would prefer to have an internal DSL for the configuration over yaml. The DSL should ofc be written in Kotlin. 😉
> What do you think?

Hey, it is a wonderful idea! Have thought the same lately when @BraisGabin mentioned that rules should be configured via properties and not the config object.

An internal rule and rule configuration dsl could bring us the 2.0 design without breaking the current one. We could prototype a new rule and rule provider class. A move from rules being visitors to registering to ast nodes can help reduce analysis time too.

Your interested in the design :)?

**Comment by BraisGabin** [comment]
I expended some time trying to move filter, aliases, active and suppressing to core last week. I didn't create any PR because I didn't find any good way to do it without breaking changes. It is possible, but the code is really complex because the API is designed for that. My idea was the same as the one proposed by Artur: Create a new API for the rules.

My ideas:
- The configuration should be injected in the rules by constructor.
- I don't like `Multirule`. It introduces a lot of complexity for little value.
- I think that with `RuleSet` and `Rule` we have enought. But we can make that a `RuleSet` can have `Rule`s and other `RuleSet`s. So we can create a tree.
- We should create a different API to add third party rules. Right now the we have the `RuleSet` API for that. But I think that they should not be realed. I mean, I should be able to add a third party `Rule` to the `style` `RuleSet`. This way we can model `RuleSet` not as an entry point of rules but just as a set of rules that can be enabled/disabed/suppressed/filter all togheter.
  - Now that I think more about this... probably the way to implement #2581 is to create "virtual sets". This way the user can create it's own aggregations of rules and add them the same "properties".

**Comment by arturbosch** [comment]
We should consider allowing easy access to `BindingContext`, `LanguageVersionSettings` and similar in the new rule api.

**Comment by BraisGabin** [comment]
We should consider removing some rules that are already handled by ktlint.
Examples:
- MaxLineLength
- NewLineAtEndOfFile
- SpacingBetweenPackageAndImports
- WildcardImport
- MandatoryBracesIfStatements

You can read a thread about this here: #2962

**Comment by schalkms** [comment]
> MandatoryBracesIfStatements

`MandatoryBracesIfStatements` actually considers more cases than `MultiLineIfElse`. 
We should be careful here when removing rules, just because they *seem* to do the same.

**Comment by BraisGabin** [comment]
> > MandatoryBracesIfStatements
> 
> `MandatoryBracesIfStatements` actually considers more cases than `MultiLineIfElse`.
> We should be careful here when removing rules, just because they _seem_ to do the same.

Yes it does. I'll open today an issue to ktlint so they can fix it.

**Comment by arturbosch** [comment]
Cross-referencing https://github.com/detekt/detekt/pull/3002#issuecomment-687082930

**Comment by BraisGabin** [comment]
> > @arturbosch I have an idea.
> > I would prefer to have an internal DSL for the configuration over yaml. The DSL should ofc be written in Kotlin. 😉
> > What do you think?
> 
> Hey, it is a wonderful idea! Have thought the same lately when @BraisGabin mentioned that rules should be configured via properties and not the config object.
> 
> An internal rule and rule configuration dsl could bring us the 2.0 design without breaking the current one. We could prototype a new rule and rule provider class. A move from rules being visitors to registering to ast nodes can help reduce analysis time too.

@arturbosch could you explaint this a bit more: "A move from rules being visitors to registering to ast nodes can help reduce analysis time too."

My understanding was that if we have the visitor as a different part we could register all the desired visitors to a single KtFile and then it will run the "visiting code" once. And this way we would reduce analysis time. But I don't see how can we accomplish that using the KtFile API. I see that we use `KtFile.accept(Visitor)` that only receives one visitor.

**Comment by arturbosch** [comment]
> > > @arturbosch I have an idea.
> > > I would prefer to have an internal DSL for the configuration over yaml. The DSL should ofc be written in Kotlin. wink
> > > What do you think?
> > 
> > 
> > Hey, it is a wonderful idea! Have thought the same lately when @BraisGabin mentioned that rules should be configured via properties and not the config object.
> > An internal rule and rule configuration dsl could bring us the 2.0 design without breaking the current one. We could prototype a new rule and rule provider class. A move from rules being visitors to registering to ast nodes can help reduce analysis time too.
> 
> @arturbosch could you explaint this a bit more: "A move from rules being visitors to registering to ast nodes can help reduce analysis time too."
> 
> My understanding was that if we have the visitor as a different part we could register all the desired visitors to a single KtFile and then it will run the "visiting code" once. And this way we would reduce analysis time. But I don't see how can we accomplish that using the KtFile API. I see that we use `KtFile.accept(Visitor)` that only receives one visitor.

Hey @BraisGabin , sorry for the late response!
Many of our rules have this scheme:

```kt
class MyRule : Rule {

  fun visitFunction() {
    // do stuff only with a KtNamedFunction node type
  }
...
```

Even so we are only visiting function the whole KtFile and all it's nodes need to be traversed.
So in the end we have x-visits for y-rules (x * y).

One possible improvement would be:

```kt
val registar: Map<KCLass<KtElement>, Rule>

rules.forEach { registrar.put(it.node, it) }

for file in ktFiles {
  file.visit(object : Visitor {
    fun visitElement(elem: KtElement) {
      registrar.get(elem::class)?.forEach { it.run(element) }
    }
  })
}
```

The rules need to be immutable so we can parallelize the `file.visit`. Note this is pseudocode and should need some tweaks to get the type conversions right when calling the rules with the nodes.

**Comment by BraisGabin** [comment]
Which is the slow part? The visit or calling a lot of empty functions? The problem of your approuch are the subtypes. What do you think about this:

```kotlin
interface KtElementVisitor {
  fun visit(item: KtElement) = Unit
}
```

This way in your rule you can do something like:

```kotlin
class MyRule : KtElementVisitor {
  override visit(item: KtAnnotation) {
    // Whatever
  }
}
```

And in the core we would do something like:

```kotlin
val fastRules = rules.filter { it is KtElementVisitor }

for file in ktFiles {
  file.visit(object : Visitor {
    fun visitElement(elem: KtElement) {
      rules.forEach { it.visit(element) }
    }
  })
}
```

As in your example, this is not final code, is just to illustrate the concept.

**Comment by chao2zhang** [comment]
I am working in a team of around 100~ Android developers. I recevied several complaints about autoCorrect is not working for all rules. It seems that autoCorrect is only working for ktlint rules, but I believe a fairly large number of rules under `styles` are actually autoCorrectible.

Given that we have not written autoCorrect rules in detekt, I couldn't safely assert that autoCorrect support is backward compatible. Therefore I can create a new issue and assign it myself to work on it.

**Comment by schalkms** [comment]
Thanks for the good detailed feedback! There’s already an open issue. Please take a look at #2560 .

**Comment by arturbosch** [comment]
As discussed in https://github.com/detekt/detekt/issues/3274 our severity model is way to complex and brings little benefit.
In https://github.com/detekt/detekt/pull/3310 we moved to an additional simpler model for the output report formats.

For 2.0 we should remove the old `Severity` class and the issue weights concept (https://github.com/detekt/detekt/issues/2272).
`SeverityLevel` of error (and warning ?)  will contribute to `maxIssues`. Then a property `build>warningsAsErrors` may be useful.
This will bring detekt's fail behavior nearer to tools like eslint.

**Comment by chao2zhang** [comment]
I would recommend us to covert this to a discussion. 

There are a few structures annotated with `@Deprecated` and `@NotApiButProbablyUsedByUsers`, which we should also remove in Detekt 2.0
